### PR TITLE
🚨 Don't use cache in non-generative models

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

When working on https://github.com/huggingface/transformers/pull/38635, I found that there are some models which have `past_key_values` in their signature, even though they cannot generate. The reason is that models were all copying from Bert

This PR clean it up and changes the copy statement to Align model and adds support for new attention API in all those models

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Yes, great cleanup! Ping me whenever it's ready and you want a review

**Comment by zucchini-nlp** [comment]
@Rocketknight1 ready for review! One thing to note: I didn't deprecate `past_key_values` from kwargs and simply deleted, since it wasn't used anyway. Do you think we need a deprecation cycle or raise an error when cache related kwargs are passed?

**Comment by Rocketknight1** [comment]
@zucchini-nlp I think it's okay! I really hope people weren't passing `past_key_values` to non-generative models anyway :grimacing:

File: src/transformers/models/git/modeling_git.py [hunk 1]
```diff
@@ -864,8 +866,6 @@ def forward(
         if output_hidden_states:
             encoder_states = encoder_states + (hidden_states,)
 
-        if not return_dict:
-            return tuple(v for v in [hidden_states, encoder_states, all_attentions] if v is not None)
```
Comment (Rocketknight1) [inline src/transformers/models/git/modeling_git.py:840]: I'm guessing these changes are just to support the new attention API?

**Review by Rocketknight1 — APPROVED** [review]
Looks good! It's a really nice cleanup, made one comment, but also we should definitely run slow tests for some of these models before merging :sweat_smile:

File: src/transformers/models/git/modeling_git.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/git/modeling_git.py:840]: ah yes, I added `can_return_tuple` instead of the recurring `if not return_dict`

File: src/transformers/models/align/modeling_align.py [hunk 2]
```diff
@@ -607,119 +632,42 @@ def __init__(self, config, position_embedding_type=None):
...
-            positional_embedding = self.distance_embedding(distance + self.max_position_embeddings - 1)
-            positional_embedding = positional_embedding.to(dtype=query_layer.dtype)  # fp16 compatibility
-
-            if self.position_embedding_type == "relative_key":
-                relative_position_scores = torch.einsum("bhld,lrd->bhlr", query_layer, positional_embedding)
-                attention_scores = attention_scores + relative_position_scores
-            elif self.position_embedding_type == "relative_key_query":
-                relative_position_scores_query = torch.einsum("bhld,lrd->bhlr", query_layer, positional_embedding)
-                relative_position_scores_key = torch.einsum("bhrd,lrd->bhlr", key_layer, positional_embedding)
-                attention_scores = attention_scores + relative_position_scores_query + relative_position_scores_key
```
Comment (Cyrilvallez) [inline src/transformers/models/align/modeling_align.py:680]: And here, why do we remove the positional_embedding? It does look like it is sometimes used, doesn't it?

File: src/transformers/models/align/modeling_align.py [hunk 3]
```diff
@@ -607,119 +632,42 @@ def __init__(self, config, position_embedding_type=None):
...
-
-        if is_cross_attention and past_key_value is not None:
-            # reuse k,v, cross_attentions
-            key_layer = past_key_value[0]
-            value_layer = past_key_value[1]
-            attention_mask = encoder_attention_mask
-        elif is_cross_attention:
-            key_layer = self.transpose_for_scores(self.key(encoder_hidden_states))
-            value_layer = self.transpose_for_scores(self.value(encoder_hidden_states))
-            attention_mask = encoder_attention_mask
```
Comment (Cyrilvallez) [inline src/transformers/models/align/modeling_align.py:634]: Why do we remove potential for cross attention here?

**Review by Cyrilvallez — COMMENTED** [review]
Hey @zucchini-nlp! Could you provide a bit more details about why we remove the cross attention and positional embeddings completely everywhere please?  🤗 It is not obvious to me, because at first glancr they look like they were used at least sometimes no?

File: src/transformers/models/align/modeling_align.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/align/modeling_align.py:680]: It is hardcoded to be `absolute` because the model doesn't need relative positions. This part was there only because of `Copied from Bert` 🙃 

I believe no-one was using it, similar to cache the official model ckpt isn't tuned for relative positions

File: src/transformers/models/align/modeling_align.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/align/modeling_align.py:634]: Same here, it is a simple vision tower and the base `AlignModel` doesn't pass any cross states to the layers, thus the kwargs are just danling from copy statements

**Comment by zucchini-nlp** [comment]
run-slow: align,wav2vec2,layoutlm,clap

**Comment by zucchini-nlp** [comment]
run-slow: align,wav2vec2,layoutlm,clap

**Comment by zucchini-nlp** [comment]
hey @Cyrilvallez do you have any comments to address?

File: src/transformers/models/align/modeling_align.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/models/align/modeling_align.py:634]: Ha yes indeed!

File: src/transformers/models/align/modeling_align.py (hunk 2 above)
Comment (Cyrilvallez) [reply src/transformers/models/align/modeling_align.py:680]: Looking [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/align/modeling_align.py#L835) and [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/align/modeling_align.py#L610-L612), there is still a possibility for the embeddings to be something else than `absolute` based on the config. So I just want to make extra sure all the main checkpoints use absolute? And maybe add 🚨 symbols to PR title

File: src/transformers/models/align/modeling_align.py [hunk 4]
```diff
@@ -847,64 +776,24 @@ def forward(
...
-                encoder_attention_mask,
-                cross_attn_past_key_value,
-                output_attentions,
-            )
-            attention_output = cross_attention_outputs[0]
-            outputs = outputs + cross_attention_outputs[1:-1]  # add cross attentions if we output attention weights
-
-            # add cross-attn cache to positions 3,4 of present_key_value tuple
-            cross_attn_present_key_value = cross_attention_outputs[-1]
-            present_key_value = present_key_value + cross_attn_present_key_value
```
Comment (Cyrilvallez) [inline src/transformers/models/align/modeling_align.py:882]: Crazy that we had this block while the model does not propagate `encoder_hidden_states`  🥵

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 5]
```diff
@@ -564,42 +563,9 @@ def forward(
         # get query proj
         query_states = self.q_proj(hidden_states).view(*q_input_shape).transpose(1, 2)
 
-        # get key, value proj
-        # `past_key_value[0].shape[2] == key_value_states.shape[1]`
-        # is checking that the `sequence_length` of the `past_key_value` is the same as
-        # the provided `key_value_states` to support prefix tuning
-        if (
-            is_cross_attention
```
Comment (Cyrilvallez) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:560]: Let's remove the line defining `is_cross_attention` here (and reapply modulars)

File: src/transformers/models/altclip/modeling_altclip.py [hunk 6]
```diff
@@ -1312,8 +1129,6 @@ def forward(
         position_ids: Optional[torch.Tensor] = None,
         head_mask: Optional[torch.Tensor] = None,
         inputs_embeds: Optional[torch.Tensor] = None,
-        encoder_hidden_states: Optional[torch.Tensor] = None,
-        encoder_attention_mask: Optional[torch.Tensor] = None,
```
Comment (Cyrilvallez) [inline src/transformers/models/altclip/modeling_altclip.py]: Here this is definitely breaking BC as this intermediate model is on the main init as well, and the args are passed correctly starting from it - I think it's fine to break it though, as I don't expect this class to be used directly, and even less with `encoder_hidden_states`, but let's be aware of it

File: src/transformers/models/altclip/modeling_altclip.py [hunk 7]
```diff
@@ -1231,56 +1075,28 @@ def forward(
...
-            encoder_extended_attention_mask = self.invert_attention_mask(encoder_attention_mask)
-        else:
-            encoder_extended_attention_mask = None
-
-        # Prepare head mask if needed
-        # 1.0 in head_mask indicate we keep the head
-        # attention_probs has shape bsz x n_heads x N x N
-        # input head_mask has shape [num_heads] or [num_hidden_layers x num_heads]
-        # and head_mask is converted to shape [num_hidden_layers x batch x num_heads x seq_length x seq_length]
-        head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)
```
Comment (Cyrilvallez) [inline src/transformers/models/altclip/modeling_altclip.py]: If `head_mask` is removed, let's remove it from the different method signatures as well

File: src/transformers/models/chinese_clip/modeling_chinese_clip.py [hunk 8]
```diff
@@ -257,119 +283,42 @@ def __init__(self, config, position_embedding_type=None):
         self.value = nn.Linear(config.hidden_size, self.all_head_size)
 
         self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
-        self.position_embedding_type = position_embedding_type or getattr(
-            config, "position_embedding_type", "absolute"
-        )
-        if self.position_embedding_type == "relative_key" or self.position_embedding_type == "relative_key_query":
-            self.max_position_embeddings = config.max_position_embeddings
-            self.distance_embedding = nn.Embedding(2 * config.max_position_embeddings - 1, self.attention_head_size)
```
Comment (Cyrilvallez) [inline src/transformers/models/chinese_clip/modeling_chinese_clip.py:266]: Similarly here, are you sure there are no hub configs with those embedding types?

File: src/transformers/models/bros/modeling_bros.py [hunk 9]
```diff
@@ -839,22 +767,17 @@ def forward(
...
         pooled_output = self.pooler(sequence_output) if self.pooler is not None else None
 
-        if not return_dict:
-            return (sequence_output, pooled_output) + encoder_outputs[1:]
-
         return BaseModelOutputWithPoolingAndCrossAttentions(
             last_hidden_state=sequence_output,
             pooler_output=pooled_output,
-            past_key_values=encoder_outputs.past_key_values,
+            past_key_values=None,
```
Comment (Cyrilvallez) [inline src/transformers/models/bros/modeling_bros.py]: Here we should remove it directly everywhere as well if we don't use it anyway

File: src/transformers/models/clap/modeling_clap.py [hunk 10]
```diff
@@ -1135,119 +1164,42 @@ def __init__(self, config, position_embedding_type=None):
         self.value = nn.Linear(config.hidden_size, self.all_head_size)
 
         self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
-        self.position_embedding_type = position_embedding_type or getattr(
-            config, "position_embedding_type", "absolute"
-        )
-        if self.position_embedding_type == "relative_key" or self.position_embedding_type == "relative_key_query":
-            self.max_position_embeddings = config.max_position_embeddings
-            self.distance_embedding = nn.Embedding(2 * config.max_position_embeddings - 1, self.attention_head_size)
```
Comment (Cyrilvallez) [inline src/transformers/models/clap/modeling_clap.py:1117]: Here as well there are no hub configs with it?

**Review by Cyrilvallez — COMMENTED** [review]
Hey! I do like it, but the PR has the potential to be quite breaking in different ways:

- position embedding type could be set in some configs on the hub
- some models present in the main init see themselves change directly (cross attention/head mask), and even if they are part of bigger models, they are still public classes. E.g. for `align`, there are no public classes using the `encoder_hidden_states`, so it would be straightforward to remove them everywhere as you did, but in `altclip`, they flow correctly from `AltCLIPTextModel` which is public, and only the main model `AltCLIPModel` does not propagate them (so removing them is not directly breaking when using `AltCLIPModel`, but is if using the public submodel `AltCLIPTextModel` 🥲)
- even method signature in fully internal classes can be breaking sometimes (though in this instance I wouldn't worry about it)

In general, I really like the changes because they clean-up a lot of non-sense in those modelings, but we need to be a bit wary of the potential implications here.
cc @ArthurZucker here for an opinion on whether we want to be agressive in favor of simplifications here, or if we want to do it through a deprecation cycle for public classes (but once again, even if public they are building block of the real bigger models)

**Comment by zucchini-nlp** [comment]
Yeah, I have the same question. On one side the code path should have never been used and isn't propagated from Base/Task models. On the other side, we never know if users found a way to exploit it by loading specific layers and re-using them

I can add proper deprecation if we think this is too aggressive and remove everything in the next 2-3 releases. A bunch of unused code path is just 😖

**Comment by ArthurZucker** [comment]
Very nice! In the era of unbloating, let's remove as much as we can, and redirect users with on the hub code for the ones that still need this? 
We can keep this for 1 release tho!

**Comment by zucchini-nlp** [comment]
@ArthurZucker @Cyrilvallez added `deprecate_kwarg` until the next v4.54 release in all forward calls. The failing test is not related

File: src/transformers/models/align/modeling_align.py [hunk 11]
```diff
@@ -615,96 +633,33 @@ def forward(
...
-        outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+        input_shape = hidden_states.shape[:-1]
+        hidden_shape = (*input_shape, -1, self.attention_head_size)
+
+        query_states = self.query(hidden_states).view(hidden_shape).transpose(1, 2)
+        key_states = self.key(hidden_states).view(hidden_shape).transpose(1, 2)
+        value_states = self.value(hidden_states).view(hidden_shape).transpose(1, 2)
+
+        attention_interface: Callable = eager_attention_forward
+        if self.config._attn_implementation != "eager":
```
Comment (ArthurZucker) [inline src/transformers/models/align/modeling_align.py:646]: that's god's work 🫡

File: src/transformers/models/altclip/modeling_altclip.py [hunk 12]
```diff
@@ -223,55 +223,16 @@ def forward(
...
-        if self.is_decoder:
-            # if cross_attention save Tuple(torch.Tensor, torch.Tensor) of all cross attention key/value_states.
-            # Further calls to cross_attention layer can then reuse all cross-attention
-            # key/value_states (first "if" case)
-            # if uni-directional self-attention (decoder) save Tuple(torch.Tensor, torch.Tensor) of
-            # all previous decoder key/value_states. Further calls to uni-directional self-attention
-            # can concat previous decoder key/value_states to current projected key/value_states (third "elif" case)
-            # if encoder bi-directional self-attention `past_key_value` is always `None`
-            past_key_value = (key_layer, value_layer)
+        key_layer = self.transpose_for_scores(self.key(hidden_states))
```
Comment (ArthurZucker) [inline src/transformers/models/altclip/modeling_altclip.py]: IMO we should remove transpose_for_scores as well!

**Review by ArthurZucker — APPROVED** [review]
😮‍💨 finally getting rid of this old code!
